### PR TITLE
JSON Schema Validation

### PR DESCRIPTION
**Pull request by trieloff**
Todo:
- [x] Provide JSON Schemas for `context` and `action`
- [x] Validate `context` and `action` in every step of the pipeline (when in production - using `tap` from #61)
- [x] Create TypeScript type files from JSON Schema
- [x] Use these type files from JSDoc so that VS Code autocomplete works
- [x] Create Markdown documentation from JSON Schema using https://github.com/adobe/jsonschema2md

### CODE REVIEW

**Comment by trieloff** [comment]
This will also help with #55 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: src/action.d.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (tripodsan) [inline src/action.d.ts]: can we add the LICENSE header here as well ?

Comment (trieloff) [reply src/action.d.ts]: Will do.

**Comment by trieloff** [comment]
This pull request **introduces 4 alerts** when merging a30dbcf3c1c27b2737898cb483536150705f2c5e into ca27533911fa8950c9d919af81dff69f88421155 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/hypermedia-pipeline/rev/pr-195871f5393302dc17500b8f09fb30d66557e1bc)

**new alerts:**

* 4 for Syntax error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging 8b84a1a55a9f74d7e23026dd58b91d5f092582f9 into ecf89b97a3dc9c39deacdbaf4aad83586df61420 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/hypermedia-pipeline/rev/pr-227641fab0c32ff54d9271507294536f7737a915)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging 6010f7e6f25b8f722e40b0c80f1b67d9532bd660 into ecf89b97a3dc9c39deacdbaf4aad83586df61420 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/hypermedia-pipeline/rev/pr-60f3e39fa28625f73e62e121c40efef60aed9acb)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This pull request **introduces 1 alert** when merging da1c7531ca79002a3ee268dc78a4928a93183ecc into ecf89b97a3dc9c39deacdbaf4aad83586df61420 - [view on LGTM.com](https://lgtm.com/projects/g/adobe/hypermedia-pipeline/rev/pr-9114466370c68046225163dff77ab92d7bc2fef1)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by trieloff** [comment]
This fixes #55 (at least for the data types in the pipeline), results can be seen here: https://github.com/adobe/hypermedia-pipeline/blob/json-schema-experiments/docs/README.md

File: src/pipeline.js [hunk 2]
```diff
@@ -13,13 +13,8 @@ const _ = require('lodash/fp');
 const Promise = require('bluebird');
 
 const nopLogger = {
-  error: () => {},
```
Comment (tripodsan) [inline src/pipeline.js]: why did you remove this?
we should ensure that `action.logger` always contains a valid logger

File: src/utils/validate.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const Ajv = require('ajv');
+const fs = require('fs-extra');
+const path = require('path');
+
+const ajv = new Ajv({ allErrors: true, verbose: true });
+const schemadir = path.dirname(__dirname);
+fs.readdirSync(schemadir)
```
Comment (tripodsan) [inline src/utils/validate.js]: don't do this here. otherwise it will always load (and block) when `validate.js` is required. also in production.

File: src/utils/validate.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  .filter(file => file.match(/\.schema\.json$/))
+  .map(file => ajv.addSchema(fs.readJSONSync(path.resolve(schemadir, file))));
+
```
Comment (tripodsan) [inline src/utils/validate.js]: how about:

````js
let ajv;

async getValidator() {
  if (!ajv) {
    ajv = new Ajv({ allErrors: true, verbose: true });
    (await fs.readdir(dirname))
       .filter(file => file.match(/\.schema\.json$/))
      .map(file => ajv.addSchema(fs.readJSONSync(path.resolve(schemadir, file))));
  }
 return ajv;
}
````

File: src/utils/validate.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+function validate(context, action, index) {
```
Comment (tripodsan) [inline src/utils/validate.js]: async ?

File: src/pipeline.js (hunk 2 above)
Comment (trieloff) [reply src/pipeline.js]: The `nopLogger` is only used in the Pipeline itself – and the removed methods are never called.

Comment (trieloff) [reply src/pipeline.js]: Brought the functions back, but make sure they still get coverage, even when not called.

File: src/utils/validate.js (hunk 4 above)
Comment (trieloff) [reply src/utils/validate.js]: It ended up being a bit more elaborate than that, but changed it.

File: src/utils/validate.js (hunk 5 above)
Comment (trieloff) [reply src/utils/validate.js]: I had to change the pipeline, too – but now all taps can be async
